agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1082 @@
+// it will try to renew it.
+func (c *CAManager) RenewIntermediate(isPrimary bool) error {
+	// Grab the 'lock' right away so the provider/config can't be changed out while we check
+	// the intermediate.
+	if err := c.setState(CAStateRenewIntermediate, true); err != nil {
+		return err
+	}
+	defer c.setState(CAStateReady, false)
+
+	provider, _ := c.getCAProvider()
+	if provider == nil {
+		// this happens when leadership is being revoked and this go routine will be stopped
+		return nil
+	}
+	// If this isn't the primary, make sure the CA has been initialized.
+	if !isPrimary && !c.configuredSecondaryCA() {
+		return fmt.Errorf("secondary CA is not yet configured.")
+	}
+
+	state := c.srv.fsm.State()
+	_, activeRoot, err := state.CARootActive(nil)
+	if err != nil {
+		return err
+	}
+
+	// If this is the primary, check if this is a provider that uses an intermediate cert. If
+	// it isn't, we don't need to check for a renewal.
+	if isPrimary {
+		_, config, err := state.CAConfig(nil)
+		if err != nil {
+			return err
+		}
+
+		if _, ok := ca.PrimaryIntermediateProviders[config.Provider]; !ok {
+			return nil
+		}
+	}
+
+	activeIntermediate, err := provider.ActiveIntermediate()
+	if err != nil {
+		return err
+	}
+
+	if activeIntermediate == "" {
+		return fmt.Errorf("datacenter doesn't have an active intermediate.")
+	}
+
+	intermediateCert, err := connect.ParseCert(activeIntermediate)
+	if err != nil {
+		return fmt.Errorf("error parsing active intermediate cert: %v", err)
+	}
+
+	if lessThanHalfTimePassed(time.Now(), intermediateCert.NotBefore.Add(ca.CertificateTimeDriftBuffer),
+		intermediateCert.NotAfter) {
+		return nil
+	}
+
+	// Enough time has passed, go ahead with getting a new intermediate.
+	renewalFunc := c.getIntermediateCAPrimary
+	if !isPrimary {
+		renewalFunc = c.getIntermediateCASigned
+	}
+	if err := renewalFunc(provider, activeRoot); err != nil {
+		return err
+	}
+
+	if err := c.persistNewRootAndConfig(provider, activeRoot, nil); err != nil {
+		return err
+	}
+
+	c.setCAProvider(provider, activeRoot)
+	return nil
+}
+
+// secondaryCARootWatch maintains a blocking query to the primary datacenter's
+// ConnectCA.Roots endpoint to monitor when it needs to request a new signed
+// intermediate certificate.
+func (c *CAManager) secondaryCARootWatch(ctx context.Context) error {
+	args := structs.DCSpecificRequest{
+		Datacenter: c.srv.config.PrimaryDatacenter,
+		QueryOptions: structs.QueryOptions{
+			// the maximum time the primary roots watch query can block before returning
+			MaxQueryTime: c.srv.config.MaxQueryTime,
+		},
+	}
+
+	c.logger.Debug("starting Connect CA root replication from primary datacenter", "primary", c.srv.config.PrimaryDatacenter)
+
+	retryLoopBackoff(ctx, func() error {
+		var roots structs.IndexedCARoots
+		if err := c.srv.forwardDC("ConnectCA.Roots", c.srv.config.PrimaryDatacenter, &args, &roots); err != nil {
+			return fmt.Errorf("Error retrieving the primary datacenter's roots: %v", err)
+		}
+
```

Comment:
safety:
if the wan read above takes a long time, the context may be canceled when it next wakes up so it might be beneficial to add a quick:
```
select {
  case <-ctx.Done():
    return ctx.Err()
  default:
}
```